### PR TITLE
Add design proposal to embed Libra's Move

### PR DESCRIPTION
**Pull request by garious**
#### Problem

Solana allows on-chain programs to be written in general purpose programming languages such as Rust or C, but currently doesn't implement any domain-specific languages from other chains. Those domain-specific languages operate at a higher-level of abstraction and programs written with them would likely be shorter and more portable.

#### Summary of Changes

Added a design to embed Libra's Move VM

Design for #5061

### CODE REVIEW

File: book/src/embedding-move.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+options.
+
+Until just recently, no popular blockchain offered a language that could expose
+the value of Solana's massively parallel runtime. Solidity contracts, for
+example, do not separate references to shared data from contract code, and
+therefore need to be executed serially to ensure deterministic behavior.  In
+practice we see that the most aggressively optimized EVM-based blockchains all
+seem to peak out around 1,200 TPS - a small fraction of what Solana can do. The
+Libra project, on the other hand, designed an on-chain programming language
+called Move that is more suitable for parallel execution. Like Pipeline VM,
```
Comment (rob-solana) [inline book/src/embedding-move.md]: might it be nice to have references to external things we're naming?

Comment (rob-solana) [reply book/src/embedding-move.md]: first mention of Pipeline in this should say "Solana Pipeline"?

File: book/src/embedding-move.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+advanced type system that allows these checks to be run by its bytecode
+verifier. Because Move bytecode can be verified, the cost of verification is
+paid just once, at the time the module is loaded on-chain. In Pipeline, the
+cost is paid each time a transaction crosses between modules. The difference
+is similar in spirit to the difference between a dynamically-typed language
+like Python versus a statically-typed language like Java. Solana's runtime
+allows dApps to be written in general purpose programming languages, but that
+comes with the cost of runtime checks when jumping between programs.
+
+This proposal attempts to defines a way to embed the Move VM such that:
```
Comment (rob-solana) [inline book/src/embedding-move.md]: defines -> define

File: book/src/embedding-move.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+All data accounts owned by Move modules must set their owners to the loader,
+`MOVE_PROGRAM_ID`. Since Move modules encapsulate their account data in the
+same way Solana programs encapsulate theirs, the Move module owner should be
+embedded in the account data. Pipeline VM will grant write access to the Move VM,
+and Move grants access to the module accounts.
+
+### Interacting with Solana programs
+
+To invoke instructions in non-Move programs, Solana would need to extend the
+Move VM with a `process_instruction()` system call. It would work the same as
```
Comment (rob-solana) [inline book/src/embedding-move.md]: here's the need to rename syscall ;)

Comment (garious) [reply book/src/embedding-move.md]: Yep! But `process_instruction` is more precise. A `sys` namespace would make that clearer.

File: book/src/embedding-move.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+`process_instruction()` Rust BPF programs.
```
Comment (aeyakovenko) [inline book/src/embedding-move.md]: how do we let Move observe solana account state?  A "SolanaResource" API?

Comment (garious) [reply book/src/embedding-move.md]: It should be easier than that. You can pass parameters into Move programs. The only challenge is that they're typed. Worst case we pass in `address` parameters and then look them up with something like a `load_account_data()` system call. And for that call to work, the user will have add to include the pubkey in the transaction. We currently don't support loading accounts dynamically, although I'm fairly certain we can do that deterministically.

**Comment by garious** [comment]
`checks` was the only relevant CI here.  Post-merge reviews welcome.

File: book/src/embedding-move.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+* Solana's runtime parallelism is exposed to batches of Move and non-Move
+  transactions
+
+## Proposed Solution
+
+### Move VM as a Solana loader
+
+The Move VM shall be embedded as a Solana loader under the identifier
+`MOVE_PROGRAM_ID`, so that Move modules can be marked as `executable` with the
+VM as its `owner`. This will allow modules to load module dependencies, as well
```
Comment (jackcmay) [inline book/src/embedding-move.md]: Are you thinking thsee module dependencies will be accounts as pubkeys specified in the transaction?

Comment (garious) [reply book/src/embedding-move.md]: @jackcmay I'm thinking modules can be stored in executable accounts, but that those accounts (not transactions) would include module dependencies.

**Pull request merged**